routeadvertisement-latency.go

```diff
@@ -0,0 +1,811 @@
+}
+
+type raLatency struct {
+	measurements.BaseMeasurement
+
+	metrics          sync.Map
+	latencyQuantiles []interface{}
+	normLatencies    []interface{}
+	// list of cudn and their pods advertised by this route
+	cudnSubnet map[string]cudnPods
+	// timestamp when cudn is detected on external host and later events of ping tests
+	cudnConnTimestamp sync.Map
+	// how many routes verified during export scenario, helpful for closing the export threads
+	verifiedExportRouteCount uint64
+	// how many routes verified during impport scenario, helpful for closing the import threads
+	verifiedImportRouteCount uint64
+	// netlink channel we open with the kernel for route monitoring
+	routeCh chan netlink.RouteUpdate
+	// channel to notify closure of netlink monitoring and export workers
+	exportDoneCh chan struct{}
+	// channel to notify closure of import workers
+	importDoneCh chan struct{}
+	udnclientset userdefinednetworkclientset.Interface
+	// channel wused between main thread and import threads. Main thread notifies import threads which routes to generate.
+	routeImportChan chan routeImport
+	wg              sync.WaitGroup
+}
+
+type raLatencyMeasurementFactory struct {
+	measurements.BaseMeasurementFactory
+}
+
+func NewRaLatencyMeasurementFactory(configSpec config.Spec, measurement types.Measurement, metadata map[string]interface{}) (measurements.MeasurementFactory, error) {
+	return raLatencyMeasurementFactory{
+		measurements.NewBaseMeasurementFactory(configSpec, measurement, metadata),
+	}, nil
+}
+
+func (plmf raLatencyMeasurementFactory) NewMeasurement(jobConfig *config.Job, clientSet kubernetes.Interface, restConfig *rest.Config) measurements.Measurement {
+	return &raLatency{
+		BaseMeasurement: plmf.NewBaseLatency(jobConfig, clientSet, restConfig),
+	}
+}
+
+// unlike default pod network, when a pod is created on udn network, pod ip address is retrieved from pod annotations.
+// we create list of cudn subnet and pod ip mappings. CUdn subnet is considered as a route exported to outside the cluster. When KB want's to ping test the cudn, it pings cudn's pods.
+func (r *raLatency) getPods() error {
+	var err error
+	listOptions := metav1.ListOptions{LabelSelector: fmt.Sprintf("kube-burner-uuid=%s", r.Uuid)}
+	nsList, err := r.ClientSet.CoreV1().Namespaces().List(context.TODO(), listOptions)
+	if err != nil {
+		log.Errorf("Error listing namespaces: %v", err)
+		return err
+	}
+	for _, ns := range nsList.Items {
+		podList, err := r.ClientSet.CoreV1().Pods(ns.Name).List(context.TODO(), listOptions)
+		if err != nil {
+			log.Errorf("Error listing pods in namespace %s: %v", ns.Name, err)
+			return err
+		}
+		for _, pod := range podList.Items {
+			podNetworks := make(map[string]podAnnotation)
+			ovnAnnotation, ok := pod.Annotations["k8s.ovn.org/pod-networks"]
+			if ok {
+				if err := json.Unmarshal([]byte(ovnAnnotation), &podNetworks); err != nil {
+					log.Debugf("failed to unmarshal ovn pod annotation  %v", err)
```

Comment:
Shouldn't be this message a Warn or Error?

Reply:
Agree. I will move it to Error.